airflow/models.py

```diff
@@ -2010,6 +2129,105 @@ def get(cls, key, session, deserialize_json=False):
         return v
 
 
+class XCom(Base):
+    """
+    Base class for XCom objects.
+    """
+    __tablename__ = "xcom"
+
+    id = Column(Integer, primary_key=True)
+    key = Column(String)
+    value = Column(PickleType(pickler=dill))
+    timestamp = Column(DateTime, server_default=func.current_timestamp())
+    visible_on = Column(DateTime, nullable=False)
+
+    # source information
+    task = Column(String, nullable=False)
+    dag = Column(String, nullable=False)
+
+    def __repr__(self):
+        return '<XCom "{key}" ({task} @ {visible_on})>'.format(
+            key=self.key,
+            task=self.task,
+            visible_on=self.visible_on)
+
+    @classmethod
+    @provide_session
+    def set(
+            cls,
+            key,
+            value,
+            visible_on,
+            task,
+            dag,
+            session=None):
+        """
+        Store an XCom value.
+        """
+
+        if not isinstance(task, basestring) or not isinstance(dag, basestring):
+            raise TypeError('Expected string; received {}'.format(type(task)))
+
+        session.expunge_all()
+        session.add(XCom(
+            key=key,
+            value=value,
+            visible_on=visible_on,
+            task=task,
+            dag=dag))
+        session.commit()
```

Comment:
Actually... looking at `@provide_session`, it calls `expunge_all()` *after* the function runs (line 279 of `utils.py`), meaning the XCom is removed from the session before `commit` is called. I think that's why I needed to add these lines explicitly. It looks like all the other uses of `@provide_session` are for getting data, not setting it, so this hasn't been an issue in the past.